### ISSUE TITLE
[crash] When using -x -E, the program attempts to free a string in argv

### ISSUE THREAD

**Issue by Ruanxingzhi** [Orc/discount#276]
When using -x -E, the program attempts to free a string in argv, leading to a crash.

This bug affects versions 2.x and 3.0.0a.

### Reproduce

Run:
```
./markdown -x -E 'aloha' -s '[]()'
```

Result:
```
root@worklab ~/f/d/s/crashes (main) [1]# ../markdown -x -E 'aloha' -s '[]()'
=================================================================
==17402==ERROR: AddressSanitizer: attempting free on address which was not malloc()-ed: 0x7ffc8af5737f in thread T0
    #0 0x5564e5f02d9a in free (/root/fuzz/discount/src/markdown+0xc2d9a) (BuildId: afc94b815b20ad8be244845d74b6694ea7bab3d9)
    #1 0x5564e5f3f113 in free_it /root/fuzz/discount/src/main.c:99:2
    #2 0x5564e5f66bca in printlinkyref /root/fuzz/discount/src/generate.c:634:23
    #3 0x5564e5f688d2 in linkyformat /root/fuzz/discount/src/generate.c:726:2
    #4 0x5564e5f631a4 in linkylinky /root/fuzz/discount/src/generate.c:773:12
    #5 0x5564e5f5f312 in text /root/fuzz/discount/src/generate.c:1391:14
    #6 0x5564e5f6e4fa in printblock /root/fuzz/discount/src/generate.c:1810:5
    #7 0x5564e5f6b8f3 in display /root/fuzz/discount/src/generate.c:2018:2
    #8 0x5564e5f6b535 in htmlify_paragraphs /root/fuzz/discount/src/generate.c:1872:5
    #9 0x5564e5f61575 in htmlify /root/fuzz/discount/src/generate.c:1916:5
    #10 0x5564e5f6b8e1 in display /root/fuzz/discount/src/generate.c:2014:2
    #11 0x5564e5f6b535 in htmlify_paragraphs /root/fuzz/discount/src/generate.c:1872:5
    #12 0x5564e5f61575 in htmlify /root/fuzz/discount/src/generate.c:1916:5
    #13 0x5564e5f60aba in mkd_document /root/fuzz/discount/src/generate.c:2065:6
    #14 0x5564e5f47d9d in mkd_generatehtml /root/fuzz/discount/src/mkdio.c:200:5
    #15 0x5564e5f4094d in main /root/fuzz/discount/src/main.c:376:7
    #16 0x7f2db89016c9 in __libc_start_call_main csu/../sysdeps/nptl/libc_start_call_main.h:58:16
    #17 0x7f2db8901784 in __libc_start_main csu/../csu/libc-start.c:360:3
    #18 0x5564e5e683f0 in _start (/root/fuzz/discount/src/markdown+0x283f0) (BuildId: afc94b815b20ad8be244845d74b6694ea7bab3d9)

Address 0x7ffc8af5737f is located in stack of thread T0
SUMMARY: AddressSanitizer: bad-free (/root/fuzz/discount/src/markdown+0xc2d9a) (BuildId: afc94b815b20ad8be244845d74b6694ea7bab3d9) in free
==17402==ABORTING
```

### Reason

The program attempts to free the string provided by `-E` option.

![image](https://github.com/Orc/discount/assets/16996226/99d60c36-9c85-4b57-a284-700e50a959bc)

### Possible fix

Malloc a copy of the `-E` option string, and use the copy in program, so that it can be freed.

**Comment by Orc** [comment]
Thank you!  I am not certain how I managed to bobble that, given that it also triggers my debugging malloc, but in any case it's been (awkwardly) corrected, committed to sccs, and mirrored here for your verifying enjoyment.

**Comment by Ruanxingzhi** [comment]
Thank you for the quick fix. But double-free occurs in **3.0.0b**.

Run:
```
../markdown -x -E aloha -s '[](1.png) [](2.png)'

free(): double free detected in tcache 2
fish: Job 1, '../markdown -x -E aloha -s…' terminated by signal SIGABRT (Abort)
```

Compile with ASan and `-O0`, ASan says it is a use-after-free:
```
=================================================================
==31118==ERROR: AddressSanitizer: heap-use-after-free on address 0x602000000010 at pc 0x55aedf432df9 bp 0x7ffce9811cb0 sp 0x7ffce9811ca8
READ of size 1 at 0x602000000010 thread T0
    #0 0x55aedf432df8 in Qstring /root/discount-1213/generate.c:153:5
    #1 0x55aedf436b23 in printlinkyref /root/discount-1213/generate.c:633:2
    #2 0x55aedf4389b2 in linkyformat /root/discount-1213/generate.c:726:2
    #3 0x55aedf433224 in linkylinky /root/discount-1213/generate.c:773:12
    #4 0x55aedf42f392 in text /root/discount-1213/generate.c:1391:14
    #5 0x55aedf43e5da in printblock /root/discount-1213/generate.c:1810:5
    #6 0x55aedf43b9d3 in display /root/discount-1213/generate.c:2018:2
    #7 0x55aedf43b615 in htmlify_paragraphs /root/discount-1213/generate.c:1872:5
    #8 0x55aedf4315f5 in htmlify /root/discount-1213/generate.c:1916:5
    #9 0x55aedf43b9c1 in display /root/discount-1213/generate.c:2014:2
    #10 0x55aedf43b615 in htmlify_paragraphs /root/discount-1213/generate.c:1872:5
    #11 0x55aedf4315f5 in htmlify /root/discount-1213/generate.c:1916:5
    #12 0x55aedf430b3a in mkd_document /root/discount-1213/generate.c:2065:6
    #13 0x55aedf417dbd in mkd_generatehtml /root/discount-1213/mkdio.c:200:5
    #14 0x55aedf410973 in main /root/discount-1213/main.c:381:7
    #15 0x7fc695af76c9 in __libc_start_call_main csu/../sysdeps/nptl/libc_start_call_main.h:58:16
    #16 0x7fc695af7784 in __libc_start_main csu/../csu/libc-start.c:360:3
    #17 0x55aedf3383f0 in _start (/root/discount-1213/markdown.asan+0x283f0) (BuildId: dc35641eb8d6308bf6cd1161ec1c1d2c454424cc)

0x602000000010 is located 0 bytes inside of 6-byte region [0x602000000010,0x602000000016)
freed by thread T0 here:
    #0 0x55aedf3d2d9a in free (/root/discount-1213/markdown.asan+0xc2d9a) (BuildId: dc35641eb8d6308bf6cd1161ec1c1d2c454424cc)
    #1 0x55aedf40f113 in free_it /root/discount-1213/main.c:99:2
    #2 0x55aedf436c4a in printlinkyref /root/discount-1213/generate.c:634:23
    #3 0x55aedf4389b2 in linkyformat /root/discount-1213/generate.c:726:2
    #4 0x55aedf433224 in linkylinky /root/discount-1213/generate.c:773:12
    #5 0x55aedf42f392 in text /root/discount-1213/generate.c:1391:14
    #6 0x55aedf43e5da in printblock /root/discount-1213/generate.c:1810:5
    #7 0x55aedf43b9d3 in display /root/discount-1213/generate.c:2018:2
    #8 0x55aedf43b615 in htmlify_paragraphs /root/discount-1213/generate.c:1872:5
    #9 0x55aedf4315f5 in htmlify /root/discount-1213/generate.c:1916:5
    #10 0x55aedf43b9c1 in display /root/discount-1213/generate.c:2014:2
    #11 0x55aedf43b615 in htmlify_paragraphs /root/discount-1213/generate.c:1872:5
    #12 0x55aedf4315f5 in htmlify /root/discount-1213/generate.c:1916:5
    #13 0x55aedf430b3a in mkd_document /root/discount-1213/generate.c:2065:6
    #14 0x55aedf417dbd in mkd_generatehtml /root/discount-1213/mkdio.c:200:5
    #15 0x55aedf410973 in main /root/discount-1213/main.c:381:7
    #16 0x7fc695af76c9 in __libc_start_call_main csu/../sysdeps/nptl/libc_start_call_main.h:58:16

previously allocated by thread T0 here:
    #0 0x55aedf3bce21 in strdup (/root/discount-1213/markdown.asan+0xace21) (BuildId: dc35641eb8d6308bf6cd1161ec1c1d2c454424cc)
    #1 0x55aedf4106ab in main /root/discount-1213/main.c:350:26
    #2 0x7fc695af76c9 in __libc_start_call_main csu/../sysdeps/nptl/libc_start_call_main.h:58:16

SUMMARY: AddressSanitizer: heap-use-after-free /root/discount-1213/generate.c:153:5 in Qstring
Shadow bytes around the buggy address:
  0x601ffffffd80: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x601ffffffe00: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x601ffffffe80: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x601fffffff00: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x601fffffff80: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
=>0x602000000000: fa fa[fd]fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x602000000080: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x602000000100: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x602000000180: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x602000000200: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x602000000280: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
Shadow byte legend (one shadow byte represents 8 application bytes):
  Addressable:           00
  Partially addressable: 01 02 03 04 05 06 07 
  Heap left redzone:       fa
  Freed heap region:       fd
  Stack left redzone:      f1
  Stack mid redzone:       f2
  Stack right redzone:     f3
  Stack after return:      f5
  Stack use after scope:   f8
  Global redzone:          f9
  Global init order:       f6
  Poisoned by user:        f7
  Container overflow:      fc
  Array cookie:            ac
  Intra object redzone:    bb
  ASan internal:           fe
  Left alloca redzone:     ca
  Right alloca redzone:    cb
==31118==ABORTING
```

**Comment by Ruanxingzhi** [comment]
Fixed in deca3b64b074192e33c21ae2c03e90437c604024.
